rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,129 @@
+
+### Handshake (new connection)
+
+The clients sends `\x00` (likely together with the previous data), and the server generates a _resumption token_ out of 16 random bytes, but with the highest bit of the first byte set. The server sends this resumption token, then its length-prefixed host ID (which is almost always a UUID in string form, but can also be an EC2 instance ID of varying length). The host ID of the server should be used in later reconnection attempts to ensure that the correct server is receiving the new connection, since the first connection might've been to a hostname or public address and thus might eventually resolve to a different server.
+
+### Handshake (reconnection)
+
+The client sends the previously-received _resumption token_, to which the server replies with a `\x01` to confirm that the token is for a connection that can be resumed or `\x00` if the token is unknown (or for a connection that's expired) followed by a connection close. In the latter case, the client should stop further reconnection attempts and treat the connection as closed by the remote side. The extra byte to confirm or deny the validity of the token is necessary to signal to the client that the connection should be treated as invalid - closing the connection will just trigger a new reconnection attempt.
+
+After the acknowledgement, both sides will send a count of how many bytes they have received in the connection.
+
+### Data exchange
+
+After both sides agree that the connection can proceed, data is exchanged in frames consisting of an integer count of how many bytes are being acknowledged as successfully received (since the previous acknowledgement, or since the beginning of the connection, to reduce the size of the variable-length integer), then a length-prefixed chunk of data. The size of the chunk must not exceed a size of 128KiB. If either side cannot proceed from the exchanged resume point, or if either side wants to explicitly close the connection, a value of `0xffff_ffff_ffff_ffff` should be sent (as the acknowledgement). Upon receiving such value, the connection should be considered closed by the remote side.
+
+### State
+
+Both sides need to maintain a log of data that has been generated by the application and potentially sent over the wire, but hasn't yet been acknowledged by the peer. Such a buffer should be limited in size to avoid unconstrained resource consumption - from some early tests, a buffer size of 2MiB seems to be sufficient to maintain near full bandwidth when transferring data across a single SSH channel, compared to a regular SSH connection that doesn't make use of the resumption protocol. The protocol does not mandate the use of a buffer with a fixed maximum size, so other approaches can be used in the future if need be.
```

Comment:
I would allow the 2MB to be configurable.  The buffer size needed could be dramatically different depending on the network conditions the client has.  For example I would be surprised if 2MB is sufficient for high throughput high latency connections like satellite or other long distance connections.

That said, there is good reason for security to keep this buffer as small as possible.  Ignoring DoS risks, this is a window of potentially sensitive data that can get replayed if the resume token is known.

Reply:
The underlying protocol is SSH, which has a 1MiB per-channel buffer limit (by default, and it's not configurable in x/crypto/ssh or openssh) so a fat long pipe is still going to be severely limited in throughput by its latency anyway.

The buffer size is a local thing anyway, it's not something that the remote needs to rely on or know about, so it's one of those things that we can improve later if need be - maybe a buffer size that depends on the amount of open channels? Even then, that's kind of a scary proposition in terms of managing resource consumption.

Reply:
Sounds fine to me, I agree that we can iterate on this later if needed.